### PR TITLE
JES Recover Closes #751

### LINKED ISSUE #751: Recovery support for JES PBE

### PR DESCRIPTION
**Pull request by mcovarr**
Doesn't close #751 but stabs in its general direction.  There is no persistent DB here, I replaced the DB-based approach in `KeyValueServiceActor` with a simple `Map` since the DB-based approach always failed to write due to referential integrity constraints that aren't going to be fixed (no `EXECUTION` or `WORKFLOW_EXECUTION` data for FK constraints).  Ruchi and/or I will have a follow-on PR to put a DB table behind this.

Also fixed "abort" to message the KV service instead of metadata service, removed now-unused `EXECUTION_INFO`-related methods, other assorted cleanup.

### CODE REVIEW

**Comment by geoffjentry** [comment]
Why does the PR say "closes #751" and the text says "Doesn't close #751"? 😕

**Comment by mcovarr** [comment]
@geoffjentry I was hoping the PR would get rolled up under #751 in ZenHub, but apparently this only bought me an "in review" label.  Oh well.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala [hunk 1]
```diff
@@ -28,40 +32,58 @@ case class JesJobExecutionActor(override val jobDescriptor: BackendJobDescriptor
...
-    // PBE TODO: use PartialFunction.orElse instead of this catch-all case, because Akka might use isDefinedAt over this partial function
-    case message => super.receive(message)
+    case KvPair(key, id @ Some(operationId)) if key.key == JesOperationIdKey =>
+      // Successful operation ID lookup during recover.
+      executor foreach { _ ! Recover(id)}
+    case KvKeyLookupFailed(_) =>
+      // Missed operation ID lookup during recover, fall back to execute.
+      executor foreach { _ ! Execute }
+    case KvFailure(_, e) =>
+      // Failed operation ID lookup during recover, crash and let the supervisor deal with it.
```
Comment (mcovarr) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala]: complete the Promise with a failure

File: backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala [hunk 2]
```diff
@@ -21,9 +21,11 @@ object AsyncBackendJobExecutionActor {
   private final case class FailAndStop(reason: Throwable) extends AsyncBackendJobExecutionActorMessage
   private final case class Finish(executionHandle: ExecutionHandle) extends AsyncBackendJobExecutionActorMessage
 
-  sealed trait ExecutionMode extends AsyncBackendJobExecutionActorMessage
+  sealed trait ExecutionMode extends AsyncBackendJobExecutionActorMessage {
+    def jesOperationId: Option[String] = None
+  }
   case object Execute extends ExecutionMode
-  final case class Resume(executionInfos: Map[String, Option[String]]) extends ExecutionMode
+  final case class Recover(override val jesOperationId: Option[String]) extends ExecutionMode
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala]: shouldn't be an `Option`

Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala]: obvs

File: backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala [hunk 3]
```diff
@@ -21,9 +21,11 @@ object AsyncBackendJobExecutionActor {
   private final case class FailAndStop(reason: Throwable) extends AsyncBackendJobExecutionActorMessage
   private final case class Finish(executionHandle: ExecutionHandle) extends AsyncBackendJobExecutionActorMessage
 
-  sealed trait ExecutionMode extends AsyncBackendJobExecutionActorMessage
+  sealed trait ExecutionMode extends AsyncBackendJobExecutionActorMessage {
+    def jesOperationId: Option[String] = None
```
Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala]: Why is there a `jesOperationId` in `AsyncBackendJobExecutionActor`? Seems awfully specific

File: backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala (hunk 2 above)
Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala]: Same comment as above, more or less

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala [hunk 4]
```diff
@@ -1,34 +1,7 @@
 package cromwell.services.keyvalue.impl
```
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: Since you're making hefty changes to the KV, can you go ahead and rename it to something more apropos as per the discussion last week?

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala [hunk 5]
```diff
@@ -1,34 +1,7 @@
...
-import akka.actor.ActorSystem
-import cromwell.core.ExecutionIndex._
-import cromwell.util.DatabaseUtil._
-import cromwell.core.{JobKey, WorkflowId}
 import cromwell.database.Database
-import wdl4s.Scope
-
-import scala.concurrent.{ExecutionContext, Future}
 
 trait KeyValueDatabaseAccess { this: Database =>
```
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: I'm assuming this is empty for future expansion?

**Comment by geoffjentry** [comment]
👍

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala (hunk 4 above)
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: Happy to TechTalk™️ this on Monday.  The choice of name may depend on whether this same store is used for call caching.

File: backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala (hunk 3 above)
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/async/AsyncBackendJobExecutionActor.scala]: Yeah that's not right, I need to generify that.

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala (hunk 5 above)
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: yes

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala (hunk 4 above)
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: FWIW I've got the tall day training tomorrow. 

I've been thinking about CC for most of the morning but that's had the effect of increasing the ways I could see it working, not decreasing. However using the KV seems among the less likely of the reasonable possibilities but who knows

Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueDatabaseAccess.scala]: I take that last statement back. Minus the "but who knows" portion

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala [hunk 6]
```diff
@@ -30,23 +31,13 @@ case class KeyValueServiceActor(serviceConfig: Config, globalConfig: Config) ext
     }
   }
 
-  private def doPut(put: KvPut): Future[KvResponse] = {
-    upsertExecutionInfo(
-      put.pair.key.workflowId,
-      put.pair.key.jobKey,
-      Map(put.pair.key.key -> put.pair.value)
-    ).map(_ => KvPutSuccess(put))
+  private def doPut(put: KvPut): Future[KvResponse] = Future.successful {
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: I don't like the way the Futures are structured in this Actor.

We have a receive that spins up a Future which modifies the state out-of-bound of the receive block. All sorts of race conditions are possible there...

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala [hunk 7]
```diff
@@ -145,33 +147,26 @@ class JesAsyncBackendJobExecutionActor(override val jobDescriptor: BackendJobDes
...
-              case Success(_) => jobLogger.info(s"$tag Aborted $jobId")
-              case Failure(ex) => jobLogger.warn(s"$tag Failed to abort $jobId: ${ex.getMessage}")
-            }
-          case None => jobLogger.warn(s"$tag Failed to abort ${jobDescriptor.call.fullyQualifiedName}. Run Id metadata field was empty.")
-        }
+      serviceRegistryActor ! KvGet(ScopedKey(jobDescriptor.descriptor.id, jobDescriptor.key, JesOperationIdKey))
+    case KvPair(scopedKey, operationId) if scopedKey.key == JesOperationIdKey =>
+      operationId match {
+        case Some(id) =>
+          jobLogger.info(s"$tag Aborting $id")
```
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: I thought we were using the operation ID to Restore an operation... not abort it?

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala (hunk 6 above)
Comment (Horneth) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: I don't think that a `Future.successful` executes asynchronously. It just wraps whatever code is inside into a successful Future and returns.

Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: @Horneth speaks the truth

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala [hunk 8]
```diff
@@ -28,40 +33,59 @@ case class JesJobExecutionActor(override val jobDescriptor: BackendJobDescriptor
...
+  }
+
+  override def recover: Future[BackendJobExecutionResponse] = {
+    import JesJobExecutionActor._
+
+    for {
+      _ <- launchExecutor
+      _ = serviceRegistryActor ! KvGet(ScopedKey(jobDescriptor.descriptor.id, jobDescriptor.key, JesOperationIdKey))
+      c <- completionPromise.future
+    } yield c
```
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala]: I would appreciate a tech talk (sort of like Khalid's database explanapaloosa) to explain what on earth is going on with these promises in this actor at some point.

E.g. why does a KvFailure above complete the promise with a failure but a KvSuccess does not, wouldn't the promise already have completed here? What even IS a promise? Much questions, many confusing.

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala (hunk 6 above)
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: Yup, I know how that works but look at the `receive` logic: Ignoring the specific implemention of `doPut`, the `receive` should not farm out state manipulation to something that returns a `Future`

Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: @cjllanwarne also speaks the truth

Comment (Horneth) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: Right - although I imagine this will not be a problem when it switches to a Database implementation.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala (hunk 7 above)
Comment (mcovarr) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: We use it for both actually.  The above is a somewhat unrelated refactoring of the abort logic to use the KV store and not the metadata service.

Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: As a side note, I don't know why abort even needs to look remotely for operation ID - the actor already uses the operation ID every time it queries status, so why is abort a special case?

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala (hunk 6 above)
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: Perhaps, since this is currently CR~~UD~~. But things change and I'm very wary of "it's fine to have race conditions because they're unlikely"

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala (hunk 8 above)
Comment (mcovarr) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesJobExecutionActor.scala]: I'll point out first that @kshakir seemed to be suggesting a different approach over in his PR though I'm not sure what that approach would be.  Anyway here's how the current scheme works:

Both `recover` above and `execute` below launch an executor actor and hand back a `Future` based on the completion promise.  `execute` just tells the executor to run with an `Execute` message.  `recover` is a bit more complicated, it must first try to fetch the operation ID based on the job key before messaging the executor.  So the first thing `recover` does is query the KV store for the operation ID.

The KV store operation ID query can be answered in one of three ways: 

* `KvPair(...)` which provides an operation ID and allows for true recovery.
* `KvKeyLookupFailed(...)` where there is no operation ID found.  This is less than ideal but not a disaster, the logic just falls back to running the job anew.
* `KvFailure` where something went horribly wrong.

The `Promise` is used here as a proxy to the `Future` that is required to be returned by the `execute` and `recovery` method signatures.  Once the outcome of the recovery or execution is known, that `Promise` is completed either successfully with the response, or unsuccessfully with the reason for failure.  

In the `KvFailure` case something has gone horribly wrong, so the `Future` is completed as failed.  In the other cases the execution is proceeding normally and the existing code will complete the `Future` at a later point as it did prior to these changes.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala (hunk 7 above)
Comment (mcovarr) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: Uh hmm that does sound right.  @scottfrazer?

File: services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala (hunk 6 above)
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: In general I agree w/ the sentiment that @cjllanwarne has here. I get skeeved out by stuff which are AOK now but are taking advantage of technicalities and could easily go horribly wrong in a manner which is easy to escape code review.

At the very least it would be helpful to have a comment in an obvious location stating that if this ever *was* async the whole thing should be changed.

Comment (mcovarr) [inline services/src/main/scala/cromwell/services/keyvalue/impl/KeyValueServiceActor.scala]: This is just a super short term placeholder for the PR after this one and #1200

**Comment by cjllanwarne** [comment]
As a halfway house (which I guess is what this is?): 👍
